File: python/gilded_rose.py
# -*- coding: utf-8 -*-

class GildedRose(object):

    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            self.update_quality_for_item(item)
    
    def update_quality_for_item(self, item):
        if item.name == "Sulfuras, Hand of Ragnaros":
            return
        
        if item.name != "Aged Brie" and item.name != "Backstage passes to a TAFKAL80ETC concert":
            ## Decreasing quality by 1 except brie, concert and sulfuras
            if item.quality > 0:
                item.quality = item.quality - 1
        else:
            ## Either brie or concert
            if item.quality < 50:
                item.quality = item.quality + 1
                if item.name == "Backstage passes to a TAFKAL80ETC concert":
                    ## concert will potentially enter both if conditions
                    if item.sell_in < 11:
                        if item.quality < 50:
                            item.quality = item.quality + 1
                    if item.sell_in < 6:
                        if item.quality < 50:
                            item.quality = item.quality + 1


        ## This will always be executed
        item.sell_in = item.sell_in - 1
        if item.sell_in < 0:
            ## expired
            if item.name != "Aged Brie":
                if item.name != "Backstage passes to a TAFKAL80ETC concert":
                    if item.quality > 0:
                        ## if not brie, concert or sulfuras, decrease quality by 1 more
                        item.quality = item.quality - 1
                else:
                    # concert
                    item.quality = item.quality - item.quality ## set quality to 0
            else:
                ## brie
                if item.quality < 50:
                    item.quality = item.quality + 1


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)

File: go/gildedrose/gildedrose.go
package gildedrose

import "strings"

type Item struct {
	Name            string
	SellIn, Quality int
}

type ItemUpdater interface {
	UpdateQuality(item *Item)
	Matches(item *Item) bool
}

type AgedBrieUpdater struct{}

func (u *AgedBrieUpdater) Matches(item *Item) bool {
	return item.Name == "Aged Brie"
}

func (u *AgedBrieUpdater) UpdateQuality(item *Item) {
	decreaseSellIn(item)
	if expired(item) {
		increaseQuality(item, 2)
	} else {
		increaseQuality(item, 1)
	}
}

type BackstagePassUpdater struct{}

func (u *BackstagePassUpdater) Matches(item *Item) bool {
	return strings.HasPrefix(item.Name, "Backstage passes")
}

func (u *BackstagePassUpdater) UpdateQuality(item *Item) {
	decreaseSellIn(item)
	if expired(item) {
		item.Quality = 0
	} else if item.SellIn < 5 {
		increaseQuality(item, 3)
	} else if item.SellIn < 10 {
		increaseQuality(item, 2)
	} else {
		increaseQuality(item, 1)
	}
}

type SulfurasUpdater struct{}

func (u *SulfurasUpdater) Matches(item *Item) bool {
	return item.Name == "Sulfuras, Hand of Ragnaros"
}

func (u *SulfurasUpdater) UpdateQuality(item *Item) {
}

type DefaultUpdater struct{}

func (u *DefaultUpdater) Matches(item *Item) bool {
	return true
}

func (u *DefaultUpdater) UpdateQuality(item *Item) {
	decreaseSellIn(item)
	if expired(item) {
		increaseQuality(item, -2)
	} else {
		increaseQuality(item, -1)
	}
}

type ConjuredUpdater struct{}

func (u *ConjuredUpdater) Matches(item *Item) bool {
	return strings.HasPrefix(item.Name, "Conjured")
}

func (u *ConjuredUpdater) UpdateQuality(item *Item) {
	decreaseSellIn(item)
	if expired(item) {
		increaseQuality(item, -4)
	} else {
		increaseQuality(item, -2)
	}
}

var itemUpdaters = []ItemUpdater{
	&SulfurasUpdater{},
	&AgedBrieUpdater{},
	&BackstagePassUpdater{},
	&ConjuredUpdater{},
	&DefaultUpdater{},
}

func UpdateQuality(items []*Item) {
	for _, item := range items {
		for _, updater := range itemUpdaters {
			if updater.Matches(item) {
				updater.UpdateQuality(item)
				break
			}
		}
	}
}

func increaseQuality(item *Item, increase int) {
	item.Quality = item.Quality + increase
	item.Quality = max(0, min(item.Quality, 50))
}

func expired(item *Item) bool {
	return item.SellIn < 0
}

func decreaseSellIn(item *Item) {
	item.SellIn = item.SellIn - 1
}
